pkg/ottl/parser.go

```diff
@@ -306,3 +324,70 @@ func (s *Statements[K]) Eval(ctx context.Context, tCtx K) (bool, error) {
+// and will handle errors returned by conditions based on an ErrorMode.
+// By default, the conditions are ORed together, but they can be ANDed together using the WithLogicOperation option.
+type ConditionSequence[K any] struct {
+	conditions        []*Condition[K]
+	errorMode         ErrorMode
+	telemetrySettings component.TelemetrySettings
+	logicOp           LogicOperation
+}
+
+type ConditionSequenceOption[K any] func(*ConditionSequence[K])
+
+// WithLogicOperation sets the LogicOperation of a ConditionSequence
+// When setting AND the conditions will be ANDed together.
+// When setting OR the conditions will be ORed together.
+func WithLogicOperation[K any](logicOp LogicOperation) ConditionSequenceOption[K] {
+	return func(c *ConditionSequence[K]) {
+		c.logicOp = logicOp
+	}
+}
+
+// NewConditionSequence creates a new ConditionSequence with the provided Condition slice, ErrorMode, and component.TelemetrySettings.
+// You may also augment the ConditionSequence with a slice of ConditionSequenceOption.
+func NewConditionSequence[K any](conditions []*Condition[K], errorMode ErrorMode, telemetrySettings component.TelemetrySettings, options ...ConditionSequenceOption[K]) ConditionSequence[K] {
+	s := ConditionSequence[K]{
+		conditions:        conditions,
+		errorMode:         errorMode,
+		telemetrySettings: telemetrySettings,
+		logicOp:           Or,
+	}
+	for _, op := range options {
+		op(&s)
+	}
+	return s
+}
+
+// Eval evaluates the result of each Condition in the ConditionSequence.
+// The boolean logic between conditions is based on the ConditionSequence's Logic Operator.
+// If using the default OR LogicOperation, if any Condition evaluates to true, then true is returned and if all Conditions evaluate to false, then false is returned.
+// If using the AND LogicOperation, if any Condition evaluates to false, then false is returned and if all Conditions evaluate to true, then true is returned.
+// When the ErrorMode of the ConditionSequence is `propagate`, errors cause the evaluation to be false and an error is returned.
+// When the ErrorMode of the ConditionSequence is `ignore`, errors cause the evaluation to continue to the next condition.
+func (c *ConditionSequence[K]) Eval(ctx context.Context, tCtx K) (bool, error) {
+	var atLeastOneMatch bool
+	for _, condition := range c.conditions {
+		match, err := condition.Eval(ctx, tCtx)
+		if err != nil {
+			if c.errorMode == PropagateError {
+				err = fmt.Errorf("failed to eval condition: %v, %w", condition.origText, err)
+				return false, err
+			}
+			c.telemetrySettings.Logger.Warn("failed to eval condition", zap.Error(err), zap.String("condition", condition.origText))
+			continue
+		}
+		if match {
+			if c.logicOp == Or {
+				return true, nil
+			}
+			atLeastOneMatch = true
+		}
+		if !match && c.logicOp == And {
+			return false, nil
+		}
+	}
+	return c.logicOp == And && atLeastOneMatch, nil
```

Comment:
Can you leave a comment here reiterating your explanation that this solves the case where we are using AND logic without propagating errors but had errors in some statements?

Reply:
Good idea